lib/innerRel.ml

```diff
@@ -599,6 +620,70 @@ and module Elts = MySet.Make(O) =
       end ;
       nss
 
+(*
+ *  Strongly connected compponets, Tarjan's algorithm.
+ * From R, Sedgwick's book "Algorithms".
+ *)
+    module SCC = struct
+      module NodeMap = M.ME
+
+      type state =
+        { id : int;
+          visit : int NodeMap.t;
+          stack : elt0 list; }
+
+      let rec pop_until n ns =
+        match ns with
+        | [] -> assert false
+        | m::ns ->
+           if O.compare n m = 0 then [m],ns
+           else
+             let ms,ns = pop_until n ns in
+             m::ms,ns
+
+      let zyva kont res nodes edges =
+
+        let m = M.to_map edges in
+
+        let rec dfs n (res,s as r) =
+          try
+            NodeMap.find n s.visit,r
+          with
+          | Not_found ->
+             let id = s.id in
+             let visit = NodeMap.add n id s.visit in
+             let stack = n::s.stack in
+             let min = id in
+             let id = id+1 in
+             let s = { id; visit; stack; } in
+             let min,(res,s) as r =
+               Elts.fold
+                 (fun n (m0,r) ->
+                   let m1,r = dfs n r in  Misc.min_int m0 m1,r)
+                 (M.succs n m)
+                 (min,(res,s)) in
+             let valk =
+               try NodeMap.find n s.visit with Not_found -> assert false in
+             if Misc.int_eq min valk then
+               let scc,stack = pop_until n s.stack in
+               let visit =
+                 List.fold_left
+                   (fun visit n -> NodeMap.add n max_int visit)
+                   s.visit scc in
+               let res = kont scc res in
+               min,(res,{ s with stack; visit; })
+             else r in
```

Comment:
You mean negate condition and swap the `then` and `else` clauses? Why not.

Reply:
With `Misc.int_eq` I am sure that the code will use integer comparison. With `==` it probably will too. 

But in fact, I use non-polymorphic equality operator for  code robustness: in case some type changes, I'll be warned. This some kind of reflex.